### PR TITLE
Add support for multihost provision and prepare

### PR DESCRIPTION
**Pull request by FrNecas**
This PR adds support for provisioning multiple machines and running prepare on them as defined in the spec in #878 . The main foundation has already been in place, so just some adjustments were needed. I've been testing on the following simple plan:

```
provision:
  - name: server-one
    role: server
    how: virtual
  - name: server-two
    role: server
    how: virtual
  - name: client-one
    role: client
    how: virtual
  - name: client-two
    role: client
    how: virtual
prepare:
  - how: shell
    script: echo 'Server'
    on: server-one
  - how: shell
    script: echo 'Client'
    on: client-one
execute:
    script: tmt --help
```

There is a problem I've encountered that has to deal with the specification and how `fmf` works... It uses PyYAML which uses YAML 1.1 where `on` string is understood as a boolean scalar value (`True`) when parsed... This causes the `on: server` syntax to not work (it is parsed as `True: server`) and there is nothing I can do on tmt side to fix it (hence the blocked label :( ). Perhaps we could convert fmf to ruamel.yaml as well and require YAML 1.2 where such syntax is not supported anymore and `true` must be used. Or we could change the attribute name in the spec but I am not a big fan of this, `on` is very clean and logical. @psss @lukaszachy @thrix or anyone whom it may concern - any ideas or thoughts on this? This is the reason why the above plan's prepare will be run 8 times (each script will be run on each guest (2x4)).

Edit: Turns out it may be possible to disable `on` -> `True` mapping (see https://stackoverflow.com/a/36470466 ). Looks like there is a class-level dict so modifying it inside tmt should have an effect on parsing in fmf but I don't think this should be the permanent solution or at least we should think about alternatives

There are a few more things to be done aside from the above mentioned:

- [x] Add test coverage
- [ ] Figure out how to support this even from the CLI (has to be discussed a bit more in depth).
- [x] Fix the tests (looks like my solution for podman container naming is not sufficient and can break :) )
- [x] Workaround `on` -> `True` mapping
- [x] Implement "pre-prepare" that gives guests info about other guests (e.g. roles, IP addresses)

@The-Mule I remember you mentioned that you would be willing to help. I'd appreciate any feedback but please take it with a grain of salt due to the problem described above...

### CODE REVIEW

**Comment by The-Mule** [comment]
I believe we need to have also some kind of "implicit prepare" step  where we need to let machines know roles and addresses of all other machines.  For instance, server-one machine need to know how to reach server-two, client-one and client-two and their roles. I suppose the easier way how to do this is to export list of roles to env variables, eg. export SERVER="server-one server-two"; export CLIENT="client-one client-two" and perhaps adding ip addresses of all provisioned machines to /etc/hosts.

**Comment by FrNecas** [comment]
I've just pushed a fix for the `on`-> `True` mapping so the prepare should now be run only on the correct guests. Will work on some test coverage soon. Also good point on the implicit prepare

File: tmt/steps/prepare/install.py
Comment (FrNecas) [inline tmt/steps/prepare/install.py:55]: I've just realised that this is not the greatest design. It requires the plugins to do something to have `on` applied. I will try to refactor this a bit and will see how it goes

File: tmt.spec
Comment (thrix) [inline tmt.spec:58]: if this is a dependency of `fmf` why it needs to be also here?

**Comment by FrNecas** [comment]
I've added the implicit prepare step for multihost which exports the roles and updates /etc/hosts. One more thing is probably trying to add support for CLI. I will try to put something together :)

**Comment by FrNecas** [comment]
During our discussion with @psss we agreed that it does not make sense to block this on the config merging and we can do it later since it will need to be done for all multihost steps anyway. I am going to rebase this PR to master and remove the yaml hack, lets see what it does with new fmf :)

**Comment by psss** [comment]
There seems to be a typo in the test:

    ./test.sh: line 36: rlAssertgrep: command not found

Interesting that it failed on `epel-8` only...

**Comment by FrNecas** [comment]
Interesting that it also passed locally and all checks were green before I pushed this branch again.

File: tmt/steps/provision/__init__.py
Comment (psss) [inline tmt/steps/provision/__init__.py:33]: Would be good to mention plan name to instruct user where to look.

File: tmt/steps/provision/__init__.py
Comment (psss) [inline tmt/steps/provision/__init__.py:115]: Shall we enable the extra space only when multiple guests are provisioned? It makes the single-guest output a bit untidy.

**Review by psss — CHANGES_REQUESTED** [review]
Looks very good, this is a bunch of work! Thanks a lot for working on this. Added a couple of rather minor comments and suggestions. Plus I have two more questions here:

When looking through the output it's not clear where preparations actually happen. Would it make sense to mention guest name as well when multiple guests are provisioned?

I tried to run the whole `/plan/prepare/name` plan including the execute step but it seems to fail on the reboot script cleanup:

    tmt run -v plan -n name
    ...
    rm: cannot remove '/usr/bin/rstrnt-reboot': No such file or directory

Do you think this might be related to the changes?

File: tests/multihost/prepare/test.sh
Comment (FrNecas) [reply tests/multihost/prepare/test.sh:19]: Great to know, thanks, will remove it.

File: tmt/steps/prepare/__init__.py
Comment (FrNecas) [reply tmt/steps/prepare/__init__.py:65]: It was intentional. The map is supposed to then be added to `/etc/hosts` so I wanted to follow the naming of the file. It's a bit of an inconsistency but I think it makes sense in this case to highlight it's something different

**Comment by FrNecas** [comment]
Thanks for all the suggestions. I will look at the failing test and improve the code a bit hopefully tomorrow. Github doesn't let me respond to this:

> Do you still plan changes here? Or shall we finish the _keys handling in #908 first?

The thing is, I am not sure what the best way to approach it is. One of my thoughts would be to do it in inheritance of `wake` and other similar methods but I remember that it got a bit complicated so I ended up keeping it like it is for now. But anyway it should be unified with #908. Since this is a bigger PR with some more work to do, I'd prefer to finish #908 quickly and have a clear vision of how we want to handle `_keys` and modify this PR accordingly. Maybe we could discuss this at hacking.

**Comment by FrNecas** [comment]
@psss thanks for the comments, most has been fixed. Two things are left to do:
 1) `_keys` depending on the outcome of the discussion
 2) the fail in setting up reboot script. It's weird since in a virtual machine, it works. But then in container it fails to copy even though the commands executed are completely identical to singlehost setup :/

**Comment by FrNecas** [comment]
@psss so there were quite a lot of conflicts but I got it working. Could you please take a look again? I hope nothing broke in the rebase process... However I ran into a problem with the test that I have and there seems to be a regression in fmf with tree pruning. Not sure if it has something to do with the YAML changes. Reproducer: Create 2 plans `roles` and `noroles` and run `tmt run plan -n roles`. Both of the plans will be run, whereas the previously on `roles` was run. This is the reason for the provision test failing.

I gave debugging a quick shot and the `name` is passed correctly to `fmf.tree.prune()` which returns both of the nodes.

**Comment by psss** [comment]
> Reproducer: Create 2 plans `roles` and `noroles` and run `tmt run plan -n roles`. Both of the plans will be run, whereas the previously on `roles` was run.

I don't think that previously `plan --name roles` would not match the `noroles` plan. Because the search there was always done as a regular expression, so any substring will match. There was no change in this part of the code.

Have no idea what else could cause the problem and I think it's probably not worth investigating more the original cause of the failure, going to modify the test match the desired plan only.

**Comment by psss** [comment]
Just noticed that there is no error message reported to the user if multiple `prepare` configs are present and `name` is not defined. Example config:

```yaml
provision:
    how: local
execute:
    script: tmt --help
prepare:
  - how: shell
    script: echo one
  - how: shell
    script: echo two
```

Previously the following would be reported:

    Missing 'name' in the prepare step config of the '/plans/example' plan.

Which means that both `prepare` plugins share the same name `default` and corresponding workdir.

**Comment by psss** [comment]
Now tests passed, only the `epel-8` job failed on `/tests/multihost/prepare`. Seems that the `on` keyword is not respected there.

**Comment by psss** [comment]
Problem seems to be caused by older `ruamel.yaml` which still interprets `on` as a `boolean`:

    # fmf show --name name
    /plan/prepare/name
    execute: {'script': 'tmt --help'}
    prepare: {True: 'server-one', 'how': 'shell', 'script': "echo 'Server one'"}
    provision: [{'how': 'container', 'name': 'server-one', 'role': 'server'},
    {'how': 'container', 'name': 'server-two', 'role': 'server'},
    {'how': 'container', 'name': 'client-one', 'role': 'client'},
    {'how': 'container', 'name': 'client-two', 'role': 'client'}]

    # rpm -q python3-ruamel-yaml
    python3-ruamel-yaml-0.15.41-2.el8.x86_64

Regarding the newer `ruamel.yaml` version there is no response from the maintainer:

    https://bugzilla.redhat.com/show_bug.cgi?id=1700334

Seems the only way forward for now is to handle the `True --> 'on'` conversion directly in `tmt`.

File: tmt/steps/__init__.py
Comment (psss) [inline tmt/steps/__init__.py:53]: I'm not sure about this change. This means that mutliple prepare instances can have the same name. The name is meant to be a unique identifier of the individual steps instance/config and is used for the workdir as well. For example the following `discover` step config:

```yaml
discover:
  - how: fmf
    url: https://github.com/psss/fmf
  - how: fmf
    url: https://github.com/psss/tmt
```

would fail because tests are fetched into the same directory:

```
fatal: destination path '/var/tmp/tmt/run-094/plans/example/discover/default/tests' already exists and is not an empty directory.
```

Oher steps do not use much workdir (yet), but I guess in the future they could and we should ensure there are no collisions.

**Review by psss — COMMENTED** [review]
Thanks for the changes. Looks good in general. Added one concern related to ip address versus hostname (maybe we could skip that for now) and ran into an issue when trying to execute the whole plan. Added a simple test for this scenario in a0f3b57. Seems the problem appears with `container` provision only and works fine with `virtual`. Could you please have a look?

File: tmt/steps/prepare/__init__.py
Comment (FrNecas) [reply tmt/steps/prepare/__init__.py:70]: You are right, this could be a problem. Though I am not sure what the best way to approach this is.

Comment (psss) [reply tmt/steps/prepare/__init__.py:70]: Yeah, I don't see a quick solution either. As mentioned in the overall summary, perhaps we could skip this for, just mark with a `FIXME` comment and do a proper IP address detection in a separate pull request.

**Comment by pcahyna** [comment]
Hello @psss, I understand that "`on` is very clean and logical", but is it really worth the troubles you have? If you work around it by using some hacks, the resulting format will not be standard YAML and will cease to be usable with generic YAML manipulating tools, which is not clean and logical anymore. What about using e.g. `onhost`?

**Comment by pcahyna** [comment]
> Silly idea but how about using `at` instead of `on`? https://www.differencebetween.com/difference-between-on-and-at/ Imho it can be used as `to refer to an address` use case

Or `host` - whatever, just please something that parses as a string if you intend it to be a string.

`<bikeshed>`
I usually look for prior art when trying to come up with naming. `host` would correspond to Ansible's `hosts` (list of hosts referring to an inventory, which is quite similar to what is intended here). Is there some popular tool which refers to hosts, target or something similar using a preposition?
`</bikeshed>`

**Comment by pcahyna** [comment]
... or `hosts` as it can be a list :-) I understand that using a preposition has the advantage of catching both the singular (scalar) and plural (list) cases.

**Comment by pcahyna** [comment]
after a discussion, the best solution seems to be `where` - there is already `how`. #992

**Comment by psss** [comment]
Note that #992 has been merged.

**Comment by psss** [comment]
@FrNecas, if I remember correctly there were just a few minor adjustments left after the latest review. Could you please rebase on the latest master? Would be nice to get this merged soon. Thanks.

File: tmt/steps/provision/podman.py
Comment (FrNecas) [inline tmt/steps/provision/podman.py:143]: Good catch!

Comment (psss) [reply tmt/steps/provision/podman.py:143]: Fortunately `podman` has these two nicely described in the man page.

File: tmt/steps/__init__.py
Comment (lukaszachy) [inline tmt/steps/__init__.py:280]: This can be removed, right? AFAIK attribute was renamed

Comment (psss) [reply tmt/steps/__init__.py:280]: Yes. Hm, it got somehow lost during the rebase/squash.

**Comment by psss** [comment]
Hm, one interesting thing: From the [logs](http://artifacts.dev.testing-farm.io/5b879ebe-8b6e-4932-9991-195efd82bbad/) it seems that the `order` is not properly set for the `multihost` prepare steps:

```
how: shell
order: 50
guest: client-two
overview: 1 script found
script: echo 'All'

how: multihost
summary: Setup guest for multihost testing
name: multihost
order: 50
guest: client-two
```

@FrNecas any quick idea, what might be wrong?

File: tmt/utils.py
Comment (psss) [inline tmt/utils.py:48]: With regards to our recent discussion about the most suitable default orders, @lukaszachy, @thrix, does it make sense to you to place multihost prepare step before required packages installation? Would be good to choose the best possible default here.

File: tmt/steps/prepare/multihost.py
Comment (lukaszachy) [inline tmt/steps/prepare/multihost.py:49]: This doesn't work and I can't find out why - check 'environment' in debug log - role is never set there.
I've added print on next line and self.step.plan.environment has only TMT_PLAN_DATA set

Comment (lukaszachy) [reply tmt/steps/prepare/multihost.py:49]: Also I don't see any quoting - spaces in role name might make it difficult

Comment (lukaszachy) [reply tmt/steps/prepare/multihost.py:49]: Maybe because environment is a property

Comment (lukaszachy) [reply tmt/steps/prepare/multihost.py:49]: It works with `self.step.plan._environment[role] = formatted_guests`

Comment (psss) [reply tmt/steps/prepare/multihost.py:49]: Yes, `Plan.environment` is a property. It would be better to handle this a bit more nicely to store the roles. What about storing roles separately and aggregating them as part of the property handling?

Comment (psss) [reply tmt/steps/prepare/multihost.py]: For now let's just update `Plan._environment`, we can polish later.

File: tmt/utils.py
Comment (psss) [reply tmt/utils.py]: Let's go with `10` for now, this should be reasonably early.

**Pull request merged**
